### ISSUE TITLE
Observable._deferUnregister generates synchronous behavior in BabylonNative causing event callbacks to get dropped when Observable.addOnce is used

### ISSUE THREAD

**Issue by chrisfromwork** [BabylonJS/Babylon.js#9362]
**Repro**

It's unclear whether this repros when using Babylon.js in a browser. It does repro in BabylonNative builds using babylon.max.js. To recreate:
1. Make multiple subscriptions to WebXRSessionManager.onXRSessionInit using addOnce function
2. Call WebXRSessionManager.onXRSessionInit.notifyObservers

- Expected result:
All subscriptions see event callbacks fired. addOnce subscribers are correctly removed after seeing a single event callback.

- Current result:
addOnce subscribers are synchronously removed from the Observable._observers array when still iterating over the Observable._observers array. Observable._deferUnregister effectively reduces the Observable._observers array length while iterating over the array (It's unclear whether this is expected behavior or if there is an assumption that deferred unregister should remove observers after finishing iterating over the Observable._observers array).

In babylon.max.js this causes the array length to change, prematurely terminating the for loop that iterates over the Observable._observers array. This then causes some observer callbacks to not fire.

**Desktop (please complete the following information):**
 - OS: Windows
 - Browser: BabylonNative
 - Version: BabylonNative master branch

This behavior has been observed for both x64 and ARM64 UWP flavors for windows mixed reality headsets and hololens 2.

**Comment by chrisfromwork** [comment]
@RaananW this is related to what we were seeing here:
https://github.com/BabylonJS/Babylon.js/pull/9307#discussion_r514570845

**Comment by chrisfromwork** [comment]
Happy to help drive a fix for this issue. It would be great if Babylon.js maintainers could call out what the expected behavior is for Observable._deferUnregister

**Comment by deltakosh** [comment]
Do you repro outside of native?

**Comment by deltakosh** [comment]
> addOnce subscribers are synchronously removed from the Observable._observers array when still iterating over the Observable._observers array. Observable._deferUnregister effectively reduces the Observable._observers array length while iterating over the array

This is the expected behavior

**Comment by deltakosh** [comment]
So I'm answering myself here: This is an issue related to babylonNative only I guess :)

**Comment by chrisfromwork** [comment]
I'm not super familiar with how setTimeout(0) is supposed to work. It seems like this has synchronous or parallel execution in BabylonNative for UWP/Chakra. I can also try testing this in Babylon.js. Would we expect different browsers to have different implementations/behaviors for setTimeout(0)?

**Comment by deltakosh** [comment]
So I've just checked on browsers and it works as expected (ie. it executes the code AFTER the loop)
@bghgary will know more but I feel like setTimeout was an issue at some point with BJN

**Comment by deltakosh** [comment]
Adding @syntheticmagus as well

**Comment by deltakosh** [comment]
Yes I feel it will be the case but just waiting for Gary or Justin to confirm ;)
